lib/crypto/c_src/pbkdf2_hmac.c

```diff
@@ -65,6 +62,30 @@ ERL_NIF_TERM pbkdf2_hmac_nif(ErlNifEnv* env, int argc,
     }
 
     return enif_make_binary(env, &out);
+}
+
+ERL_NIF_TERM pbkdf2_hmac_nif(ErlNifEnv* env, int argc,
+                             const ERL_NIF_TERM argv[])
+{
+#ifdef HAS_PKCS5_PBKDF2_HMAC
+    ErlNifUInt64 iter;
+
+    if (!enif_get_uint64(env, argv[3], &iter))
+        return EXCP_BADARG_N(env, 3, "Not integer");
+
+    if (iter < 1) {
+        return EXCP_BADARG_N(env, 3, "Must be > 0");
+    }
+
+    if (iter < 100) {
+        /* If iteration count is low enough call the NIF directly */
+        return pbkdf2_hmac(env, argc, argv);
+    }
```

Comment:
Should we do a similar check for keylen? Even with < 100 iterations, an abnormally large keylen would still block the schedulers for quite a bit.

```
> timer:tc(fun() -> binary:encode_hex(crypto:pbkdf2_hmac(sha256, <<"pass">>, <<"salt">>, 99, 1 bsl 20)), ok end).
{2235296,ok}
```

In most cases keylen would be less or equal to the size of the hmac hash used. The highest current is 64 bytes for the sha512.

```
> crypto:hash_info(sha512).
#{block_size => 128,size => 64,type => 674}
```

So perhaps it should be? :

```c
if (iter < 100 && keylen <= 64) { ... call directly ...
```

Reply:
Added keylen to the check as well.

Reply:
or maybe something like
`if (iter * keylen < DIRTY_WORK_LIMIT)`
would make sense.

Reply:
> or maybe something like `if (iter * keylen < DIRTY_WORK_LIMIT)` would make sense.

that's an interesting idea but they don't seem to scale uniformly in general, in other words `large_iter, small_keylen` has quite different characteristics than `small_iter, large_keylen` so may make it harder to reason about:

```
30> timer:tc(fun() -> binary:encode_hex(crypto:pbkdf2_hmac(sha256, <<"pass">>, <<"salt">>, 2000000, 2)), ok end).
{921226,ok}
31> timer:tc(fun() -> binary:encode_hex(crypto:pbkdf2_hmac(sha256, <<"pass">>, <<"salt">>, 200000, 20)), ok end).
{115641,ok}
32> timer:tc(fun() -> binary:encode_hex(crypto:pbkdf2_hmac(sha256, <<"pass">>, <<"salt">>, 20000, 200)), ok end).
{87254,ok}
33> timer:tc(fun() -> binary:encode_hex(crypto:pbkdf2_hmac(sha256, <<"pass">>, <<"salt">>, 2000, 2000)), ok end).
{80732,ok}
34> timer:tc(fun() -> binary:encode_hex(crypto:pbkdf2_hmac(sha256, <<"pass">>, <<"salt">>, 200, 20000)), ok end).
{83472,ok}
35> timer:tc(fun() -> binary:encode_hex(crypto:pbkdf2_hmac(sha256, <<"pass">>, <<"salt">>, 20, 200000)), ok end).
{79334,ok}
36> timer:tc(fun() -> binary:encode_hex(crypto:pbkdf2_hmac(sha256, <<"pass">>, <<"salt">>, 2, 2000000)), ok end).
{106003,ok}
```